/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package privacy_sandbox.bidding_auction_servers.bidding_service;

// Specify how to fetch the code blob.
enum FetchMode {
   // Fetch a single blob from an arbitrary url.
   FETCH_MODE_URL = 0;
   // Fetch all blobs from a specified bucket
   // and rely on a default blob specification.
   FETCH_MODE_BUCKET = 1;
   // Fetch a blob from a local path. Dev-only.
   FETCH_MODE_LOCAL = 2;
}

message BuyerCodeFetchConfig {

   // The javascript generateBid script.
   string bidding_js_path = 1;

   // URL endpoint for fetching AdTech code blob, js file.
   string bidding_js_url = 2;

   // URL endpoint to the wasm file. AdTechs can choose to include wasm file
   // which will be combined with the js file before loading into Roma.
   string bidding_wasm_helper_url = 3;

   // Period of how often to fetch AdTech code blob from the URL endpoint.
   int64 url_fetch_period_ms = 4;

   // A time out limit for HttpsFetcherAsync client to stop executing FetchUrl.
   int64 url_fetch_timeout_ms = 5;

   // Allow buyer debug URL generation.
   bool enable_buyer_debug_url_generation = 6;

   // Allow handling of console.logs from AdTech script execution.
   bool enable_adtech_code_logging = 8;

   // URL endpoint for fetching AdTech code blob, js file for protected app
   // signals.
   string protected_app_signals_bidding_js_url = 9;

   // URL endpoint to the wasm file. AdTechs can choose to include wasm file
   // which will be combined with the js file before loading into Roma.
   string protected_app_signals_bidding_wasm_helper_url = 10;

   // URL endpoint for fetching AdTech code blob, js file.
   string prepare_data_for_ads_retrieval_js_url = 11;

   // URL endpoint to the wasm file. AdTechs can choose to include wasm file
   // which will be combined with the js file before loading into Roma.
   string prepare_data_for_ads_retrieval_wasm_helper_url = 12;

   // The name of a bucket from which to fetch protected auction code blobs.
   // All blobs will be fetched from this bucket.
   // TODO: the concept of a separate wasm_helper blob is not yet supported if using
   // bucket-based fetching.
   string protected_auction_bidding_js_bucket = 14;

   // The name of the bucket's default protected auction code blob to use.
   // The default will be used if the bid request does not specify a version.
   string protected_auction_bidding_js_bucket_default_blob = 15;

   // The name of a bucket from which to fetch protected app signal code blobs.
   // TODO: the concept of a separate wasm_helper blob is not yet supported if using
   // bucket-based fetching.
   string protected_app_signals_bidding_js_bucket = 16;

   // The name of the bucket's default protected app signal code blob to use.
   // The default will be used if the bid request does not specify a version.
   string protected_app_signals_bidding_js_bucket_default_blob = 17;

   // The name of a bucket from which to fetch prepareDataForAdsRetrieval js.
   // TODO: the concept of a separate wasm_helper blob is not yet supported if using
   // bucket-based fetching.
   string ads_retrieval_js_bucket = 18;

   // The name of the bucket's default prepareDataForAdsRetrieval code blob to use.
   // The default will be used if the bid request does not specify a version.
   string ads_retrieval_bucket_default_blob = 19;

   FetchMode fetch_mode = 20;

}
